flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedStateBackend.java

```diff
@@ -137,10 +137,25 @@
      */
     boolean deregisterKeySelectionListener(KeySelectionListener<K> listener);
 
+    @Deprecated
     default boolean isStateImmutableInStateBackend(CheckpointType checkpointOptions) {
         return false;
     }
 
+    /**
+     * Whether it's safe to reuse key-value from the state-backend, e.g for the purpose of
+     * optimization.
+     *
+     * <p>NOTE: this method should not be used to check for {@link InternalPriorityQueue}, as the
+     * priority queue could be stored on different locations, e.g RocksDB state-backend could store
+     * that on JVM heap if configuring HEAP as the time-service factory.
+     *
+     * @return returns ture if safe to reuse the key-value from the state-backend.
+     */
+    default boolean isSafeToReuseState() {
```

Comment:
With this change, there are:
- `KeyedStateBackend.isSafeToReuseState`
- `AbstractKeyedStateBackend.requiresLegacySynchronousTimerSnapshots`

which I think serve the same purpose but for KV and PQ states.

How about  having them both in `KeyedStateBackend` and name accordingly (KV and PQ for example)? I think this would make API more symmetric and explicit.

Reply:
I think these two methods serveing for different purposes. The method of `isSafeToReuseState` is used to hint for SQL like operators to decide whether they need to copy the state data from the state-backend. This actually focus on the normal state access and has no relationship with checkpoint or timers.
On the other hand, `requiresLegacySynchronousTimerSnapshots` is only used to represent that whether we should use raw checkpoint stream to backup the timers. As you can see, this method serves for checkpoint and timers. If we could drop the support HEAP timers in RocksDB, we could remove `requiresLegacySynchronousTimerSnapshots` but not the `isSafeToReuseState` method.

Maybe we could rename the method `KeyedStateBackend.isSafeToReuseState` to `KeyedStateBackend.isSafeToReuseKVState`, however, I don't think `KeyedStateBackend.isSafeToReusePQState` could meet the purpose of current `AbstractKeyedStateBackend.requiresLegacySynchronousTimerSnapshots
`.

Reply:
Makes sense.

> Maybe we could rename the method KeyedStateBackend.isSafeToReuseState to KeyedStateBackend.isSafeToReuseKVState

I think this renaming alone still makes sense.